docs/auth/oidc.md

```diff
@@ -0,0 +1,231 @@
+
+### The API reference
+
+An API reference exist for the sake of **Dependency Injection**, check [Utility APIs][4]
+for extended explanation.
+
+In this OIDC example, we'll create the API reference directly in the
+`packages/app/src/apis.ts` file, it is not a requirement to put the reference in this
+file. Any location will do as long as it's available to be imported to where the API
+factory is, as well as easily accessible to the rest of the application so any package
+and plugin can inject the API instance when necessary.
+
+An example of such would be when you use an auth provider from a library installed with
+NPM, or any other library repository, you would import the API ref from the library.
+
+```ts
+export const azureOIDCAuthApiRef: ApiRef<
+  OpenIdConnectApi & ProfileInfoApi & BackstageIdentityApi & SessionApi
+> = createApiRef({
+  id: 'core.auth.azureOIDC',
+});
+```
+
+Please note a few things, the ID can be anything you want as long as it doesn't conflict
+with other refs, also we're exporting this reference, as well as the `typings`, we need to
+be able to import this reference anywhere in the app, and the `typings` will tell typescript
+what instance we're getting from DI when injecting the API. In this case we are defining
+an API for authentication, so we tell TS that this instance complies with 4 API
+interfaces:
+
+- The OICD API that will handle authentication.
+- Profile API for requesting user profile info from the auth provider in question.
+- Backstage identity API to handle and associate the user profile with backstage identity.
+- Session API, to handle the session the user will have while logged in.
+
+### The API Factory
+
+A factory is a function that can take some parameters or dependencies and return an
+instance of something, in our case it will be a function that requests some backstage
+APIs and use them to create an instance of an OIDC provider.
+
+Please note that this function only runs (creates the instance) when somewhere else in
+the app you request the DI to give you an instance of the OIDC provider using the API ref
+defined above, and the DI will only run this function the first time, from then on any
+other DI injection will just receive the same instance created the first time, basically
+the instance is cached by the DI library, a singleton.
+
+Let's add our OIDC factory to the APIs array in the `packages/app/src/apis.ts` file:
+
+```diff
++ import { OAuth2 } from '@backstage/core-app-api';
+
+export const apis: AnyApiFactory[] = [
++   createApiFactory({
++    api: azureOIDCAuthApiRef,
++    deps: {
++      discoveryApi: discoveryApiRef,
++      oauthRequestApi: oauthRequestApiRef,
++      configApi: configApiRef,
++    },
++    factory: ({ discoveryApi, oauthRequestApi, configApi }) =>
++      OAuth2.create({
++        discoveryApi,
++        oauthRequestApi,
++        provider: {
++          id: 'oidc', // This has to be 'oidc' or OAuth2 will not use oidc protocol
```

Comment:
This ID should align with the ID in the installed provider factory, i.e. `azureOIDC` currently, although I suggest we change it to something like `'acme'`. Right now this would be hitting the `/api/auth/oidc/` provider, which is not the provider we're adding here.

Reply:
For `OAuth2` to use the OIDC protocol, this `id` has to be `oidc`, otherwise, the library will use a different protocol.

Reply:
What library? Where is that logic?

Reply:
My bad, looked again and it's not the case, will change.